### ISSUE TITLE
Label issue in customised training

### ISSUE THREAD

**Issue by abc123yuanrui** [switchablenorms/DeepFashion_Try_On#46]
When I trained the network with LIP dataset (20 labels' segmentation input), this error raised
![image](https://user-images.githubusercontent.com/37522312/94094008-de1dd000-fe61-11ea-8264-5826c62ab28e.png)
Based on https://github.com/NVlabs/SPADE/issues/57 it should be some channel problem, but I believe I had changed relevant channels and total channel numbers except noise channel. Since Lip dataset doesn't have a counterpart, I used the dress channel and it seems not work properly. Any idea to fix this issue?

**Comment by MotiBaadror** [comment]
Hi @abc123yuanrui , Have you considered the label numbering of the segmentation output. There are only 14 labels in the segmentation mask. Try printing unique value for the label you are using. It shouldn't be more than 13.
```
0 -> Background
1 -> Hair
4 -> Upclothes
5 -> Left-shoe 
6 -> Right-shoe
7 -> Noise
8 -> Pants
9 -> Left_leg
10 -> Right_leg
11 -> Left_arm
12 -> Face
13 -> Right_arm
```

try below code
```
import cv2
import numpy as np
img=cv2.imread('path to label')
print(np.unique(img))
```

**Comment by abc123yuanrui** [comment]
Hi @MotiBaadror , Thanks for your answer. I have figured the issue was caused by my label number not matching. The input segmentation I used was 21 labels (20 LIP + neck). However I set it as 20. Sorry I forgot to close this issue.

**Comment by AjayMudhai** [comment]
@abc123yuanrui @MotiBaadror  Hi, Can you please tell what is label number 7 named 'noise'? and which segmentation model you used? It would be very helpful.

**Comment by abc123yuanrui** [comment]
> @abc123yuanrui @MotiBaadror Hi, Can you please tell what is label number 7 named 'noise'? and which segmentation model you used? It would be very helpful.
Hi @AjayMudhai  The label number 7 is random mask noise for covering the cloth and image. However, no such info in customised segmentation results generated by either CIHP or Self-Correction-Human-Parsing or other LIP relevant segmentation methods. According to the author of ACGPN, you can omit channel 7 when u do customised training/testing.
I personally used Self-Correction-Human-Parsing for LIP and further processed with cp-vton-plus neck detection which contains 21 labels.

**Comment by frankiesquared** [comment]
@abc123yuanrui @AjayMudhai @MotiBaadror @Mushahid2521 @ChongjianGE @taliegemen @ecarist
How do I generate a file like those in the folder /DeepFashion_Try_On/Data_preprocessing/test_label ?
I followed your directions to use https://github.com/PeikeLi/Self-Correction-Human-Parsing trained on LIP dataset and got below?

![Leather192_256Label](https://user-images.githubusercontent.com/59422631/96322029-63615400-0fcc-11eb-86e4-034910a92f8d.png)

The files in /DeepFashion_Try_On/Data_preprocessing/test_label by default look black with a silhouette.

![000001_0](https://user-images.githubusercontent.com/59422631/96322006-488edf80-0fcc-11eb-9ac3-dfca6258431c.png)


Thanks

**Comment by AjayMudhai** [comment]
@frankiesquared Hi, u jst need to save the image before applying palette.Make below shown changes in simple_extractor.py and u will get ur results.
`with torch.no_grad():
    for idx, batch in enumerate(tqdm(dataloader)):
        image, meta = batch
        img_name = meta['name'][0]
        print(img_name)
        save_name = meta['name'][0].rsplit(".",1)[0]+"_output_label_only.png"
        print(save_name)
        c = meta['center'].numpy()[0]
        s = meta['scale'].numpy()[0]
        w = meta['width'].numpy()[0]
        h = meta['height'].numpy()[0]

        output = model(image.cuda())
        upsample = torch.nn.Upsample(size=input_size, mode='bilinear', align_corners=True)
        upsample_output = upsample(output[0][-1][0].unsqueeze(0))
        upsample_output = upsample_output.squeeze()
        upsample_output = upsample_output.permute(1, 2, 0)  # CHW -> HWC
        logits_result = transform_logits(upsample_output.data.cpu().numpy(), c, s, w, h, input_size=input_size)
        parsing_result = np.argmax(logits_result, axis=2)
        parsing_result_path = os.path.join('outputs', img_name[:-4] + '.png')
        output_img = Image.fromarray(np.asarray(parsing_result, dtype=np.uint8))
        output_img.save(save_name)`
